### DISCUSSION TITLE
beyondstorage: why we failed

### DISCUSSION THREAD

**Discussion by giscus** [Xuanwo/blog#402] [Announcements]
# beyondstorage: why we failed

An infrastructure engineer, focused on distributed storage system

https://xuanwo.io/2023/01-beyond-storage-why-we-failed/

**Comment by STRRL** [comment]
🤔 边看边学

**Comment by bxb100** [comment]
还有一个原因: Rust 社区是真的好

**Comment by noirgif** [comment]
老实说看到一个项目倒下总是会感到一点伤感，带领项目的老涡的痛楚一定更加彻骨吧。能痛定思痛写下这篇post-mortem非常感激。

我读了 RethinkDB 那篇，很受启发——worse is better 和 worse is worse 两篇文章也很有意思。Worse is Better的口号过于有冲击力，我很害怕被带偏，不能很好驾驭其内涵。

站在上面的角度来讲，“盲目扩张项目”“过早分裂项目”“对细节过于推崇”这三点我很能理解——它们都分散了注意力，延迟了产品上线。Worse is Better里提到应该先目标定在做出来一个50% correct的东西（我得时时提醒自己50%算是夸张了，目的是先做出来一个符合使用场景，大多数情况下能用的东西），RethinkDB也提到追求细节而丧失先机。

“过度社区运营” 怎么影响了项目、有多大影响（是花了很多人力，还是打乱了开发节奏之类的），我不清楚具体情况，不过之前老涡写过关于TDEngine抓虫的事情，底下也有很多评论，所以如果经历了那些还是这么肯定的话，一定是有所把握了。但是为了拓展用户做一定的努力也是值得评价的。 MongoDB有个用户论坛；OpenDAL有个discord，而且加入了Apache Incubator，我觉得这都是有用的社区运营。

关于拓展用户另外一点，就像RethinkDB里说的，“用户想要的是‘a good way to do X’”。把这些对用户来说最重要的东西放在最显眼的地方就能抓住人心。我觉得最能体现 OpenDAL 是 a good way to do X 的地方就是它的主页和引导文档了。我有这三个窍门：
1. 在 Get Started 里直接地告诉用户怎么做X，比如学[S3的rust文档](https://docs.aws.amazon.com/sdk-for-rust/latest/dg/rust_s3_code_examples.html)的例子，直接告诉用户怎么操作。Rust文档因为代码结构这些例子可能只能放在 services 的文档下。这足够显眼吗？如果不，应该放在哪里？
2. 跟现有方法对比。比如和上面的S3文档对比，突出OpenDAL如何简洁的做创建、访问、删除。跟几个library访问的例子横向比较，突出OpenDAL怎么统一了这些界面。
3. <s>想不出来了</s>关于efficiently部分，一些benchmark也可以放在显眼的地方，让用户不用担心性能问题地使用。
